### PR TITLE
Feature vsan

### PR DESCRIPTION
**Pull request by YuxinZhou**
# Changes made
- Made perf entities configurable for collection. Got default entities.
- Made vSAN data as new resource, so the configuration for vSAN stays consistent to other resources. 
- Collected vSAN health metrics
- Collected vSAN storage metrics 
- Collected vSAN resynchronization metrics 
- Added additional tags
- Added unit tests for the functions of cmmds and tags.

# Configuration 
Here is a sample configuration file.
```
[[inputs.vsphere]]
  interval = "600s"
  vcenters = [ "https://10.0.0.1/sdk" ]
  username = "username"
  password = "password"

  insecure_skip_verify = true
  force_discover_on_init = true

  # Exclude all other metrics
  vm_metric_exclude = ["*"]
  datastore_metric_exclude = ["*"]
  datacenter_metric_exclude = ["*"]
  host_metric_exclude = ["*"]
  cluster_metric_exclude = ["*"]
  # Enable vSan
  vsan_enabled = true
  # By default all supported entity will be included
  vsan_perf_metric_include = ["host-domclient", "cache-disk", "disk-group", "capacity-disk"]
  # vsan_perf_metric_exclude = ["*"]

  collect_concurrency = 5
  discover_concurrency = 5
```
By vsan_enabled = true, all vSAN metrics are collected by default. These include vSAN overall health, vSAN capacity and vSAN performance metrics for supported entity type. Since the supported entity types vary across releases, a list of supported entity will be fetched through GetSupportedEntityTypes API. To include or exclude performance entity type, vsan_perf_metric_include and vsan_perf_metric_exclude can be specified. However, notice that anything not in supported entity list will be filtered out.

The vSAN performance data is available every 5 minutes by default, so a recommend interval choice for the agent is between 300s to 3600s. The collector will query performance metrics for a duration from the time last sample collected to now. If the last sample time is unavailable or expires (timestamps expire in one hour), the default collecting duration is 15 minutes.

# Sample output

```
vsphere_cluster_vsan_performance_host-domclient,clustername=Example-VSAN,dcname=Example-DC,host=host.example.com,hostname=DC0_C0_H0,moid=domain-c7,source=Example-VSAN,vcenter=localhost:8898 latencyAvg=1307,congestion=0,oio=12,iopsWrite=19,latencyMaxUnmap=0,latencyStddev=1483,latencyAvgWrite=1709,writeCount=5722,clientCacheHitRate=0,writeCongestion=0,throughputUnmap=0,throughput=313682,ioCount=8298,latencyAvgRead=415,readCount=2576,clientCacheHits=0,numOio=107657,latencyMaxRead=10780,unmapCount=0,iopsRead=8,throughputRead=36561,throughputWrite=277121,readCongestion=0,unmapCongestion=0,iops=27,iopsUnmap=0,latencyAvgUnmap=0,latencyMaxWrite=13771 1557336300000000000
vsphere_cluster_vsan_performance_cluster-domcompmgr,clustername=Example-VSAN,dcname=Example-DC,host=host.example.com,moid=domain-c7,source=Example-VSAN,uuid=XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXX,vcenter=localhost:8898 latencyAvgRead=865,iopsRecWrite=0,oio=40,tputResyncRead=0,latAvgResyncRead=0,throughputRead=13116039,latencyAvgWrite=659,throughputRecWrite=0,congestion=0,latencyAvgRecWrite=0,iopsResyncRead=0,iopsRead=434,iopsWrite=444,throughputWrite=12907193 1557336300000000000
vsphere_cluster_vsan_performance_cluster-domclient,clustername=Example-VSAN,dcname=Example-DC,host=host.example.com,moid=domain-c7,source=Example-VSAN,uuid=XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXX,vcenter=localhost:8898 iopsWrite=137,throughputWrite=6415204,latencyAvgWrite=7689,congestion=0,oio=238,iopsRead=13,throughputRead=442438,latencyAvgRead=543 1557336300000000000
vsphere_cluster_vsan_summary,clustername=Example-VSAN,dcname=Example-DC,host=host.example.com,moid=domain-c7,source=Example-VSAN,vcenter=localhost:8898 TotalBytesToSync=0i,TotalObjectsToSync=0i,TotalRecoveryETA=0i 1557336663000000000
vsphere_cluster_vsan_summary,clustername=Example-VSAN,dcname=Example-DC,host=host.example.com,moid=domain-c7,source=Example-VSAN,vcenter=localhost:8898 OverallHealth=1i 1557336663000000000
vsphere_cluster_vsan_summary,clustername=Example-VSAN,dcname=Example-DC,host=host.example.com,moid=domain-c7,source=Example-VSAN,vcenter=localhost:8898 TotalCapacityB=45129177395200i,FreeCapacityB=21377606853416i 1557336663000000000
```

### CODE REVIEW

File: plugins/inputs/vsphere/endpoint.go [hunk 1]
```diff
@@ -384,7 +399,7 @@ func (e *Endpoint) discover(ctx context.Context) error {
 	// Populate resource objects, and endpoint instance info.
 	newObjects := make(map[string]objectMap)
 	for k, res := range e.resourceKinds {
-		log.Printf("D! [inputs.vsphere]  Discovering resources for %s", res.name)
+		log.Printf("D! [inputs.vsphere] Discovering resources for %s", res.name)
```
Comment (vikramraman) [inline plugins/inputs/vsphere/endpoint.go]: Add `:` after `[inputs.vsphere]` for consistency with the above change?

**Comment by vikramraman** [comment]
Remove the `vsan-sdk` files. They should be vendored using dep. See the top level `Gopkg.toml` file.

File: plugins/inputs/vsphere/endpoint.go (hunk 1 above)
Comment (YuxinZhou) [reply plugins/inputs/vsphere/endpoint.go]: Yep, thanks.

**Comment by YuxinZhou** [comment]
Hey, I just moved vSAN go bindings to Govmomi, and use the #revision for constraint in the Gopkg.toml. Does it look good?

**Comment by prydin** [comment]
Great work on this PR! I approve.

**Pull request merged**